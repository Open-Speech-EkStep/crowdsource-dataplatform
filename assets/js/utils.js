const { HOUR_IN_SECONDS, SIXTY } = require("./constants");

const ALL_LANGUAGES = [
    {value: "Assamese",id: "as", text: "অসমীয়া", hasLocaleText: true},
    {value: "Bengali", id: "bn", text: "বাংলা", hasLocaleText: true},
    {value: "English", id: "en", text: "English", hasLocaleText: true},
    {value: "Gujarati", id: "gu", text: "ગુજરાતી", hasLocaleText: true},
    {value: "Hindi", id: "hi", text: "हिंदी", hasLocaleText: true},
    {value: "Kannada", id: "kn", text: "ಕನ್ನಡ", hasLocaleText: true},
    {value: "Malayalam", id: "ml", text: "മലയാളം", hasLocaleText: true},
    {value: "Marathi", id: "mr", text: "मराठी", hasLocaleText: true},
    {value: "Odia", id: "or", text: "ଓଡିଆ", hasLocaleText: true},
    {value: "Punjabi", id: "pa", text: "ਪੰਜਾਬੀ", hasLocaleText: true},
    {value: "Tamil", id: "ta", text: "தமிழ்", hasLocaleText: true},
    {value: "Telugu", id: "te", text: "తెలుగు", hasLocaleText: true}];

function convertPXToVH(px) {
    return px * (100 / document.documentElement.clientHeight);
}

function setPageContentHeight() {
    const $footer = $('footer');
    const $nav = $('.navbar');
    const edgeHeightInPixel = $footer.outerHeight() + $nav.outerHeight()
    const contentHeightInVH = 100 - convertPXToVH(edgeHeightInPixel)
    $('#content-wrapper').css('min-height', contentHeightInVH + 'vh');
}

function toggleFooterPosition() {
    const $footer = $('footer');
    $footer.toggleClass('fixed-bottom')
    $footer.toggleClass('bottom')
}

function fetchLocationInfo() {
    //https://api.ipify.org/?format=json
    return fetch('https://www.cloudflare.com/cdn-cgi/trace').then(res => res.text()).then(ipAddressText => {
        const dataArray = ipAddressText.split('\n');
        let ipAddress = "";
        for (let ind in dataArray) {
            if (dataArray[ind].startsWith("ip=")) {
                ipAddress = dataArray[ind].replace('ip=', '');
                break;
            }
        }
        if (ipAddress.length !== 0) {
            return fetch(`/location-info?ip=${ipAddress}`);
        } else {
            return new Promise((resolve, reject) => {
                reject("Ip Address not available")
            })
        }
    });
}

const performAPIRequest = (url) => {
    return fetch(url).then((data) => {
        if(!data.ok) {
            throw Error(data.statusText || 'HTTP error');
        } else {
            return Promise.resolve(data.json());
        }
    });
}

const getLocaleString = function() {
    performAPIRequest('/get-locale-strings')
        .then((response) => {
            localStorage.setItem('localeString', JSON.stringify(response));
        });
}

const updateLocaleLanguagesDropdown = (language) => {
    const dropDown = $('#localisation_dropdown');
    const localeLang = ALL_LANGUAGES.find(ele => ele.value === language);
    if (language.toLowerCase() === "english" || localeLang.hasLocaleText === false) {
        dropDown.html('<a id="english" class="dropdown-item" href="/changeLocale/en">English</a>');
    } else {
        dropDown.html(`<a id="english" class="dropdown-item" href="/changeLocale/en">English</a>
        <a id=${localeLang.value} class="dropdown-item" href="/changeLocale/${localeLang.id}">${localeLang.text}</a>`);
    }
}

const calculateTime = function (totalSeconds, isSeconds = true) {
    const hours = Math.floor(totalSeconds / HOUR_IN_SECONDS);
    const remainingAfterHours = totalSeconds % HOUR_IN_SECONDS;
    const minutes = Math.floor(remainingAfterHours / SIXTY);
    const seconds = Math.round(remainingAfterHours % SIXTY);
    if (isSeconds) {
        return { hours, minutes, seconds };
    } else {
        return { hours, minutes };
    }
};

const formatTime = function (hours, minutes = 0, seconds = 0) {
    let result = "";
    if (hours > 0) {
        result += `${hours} hrs `;
    }
    if (minutes > 0) {
        result += `${minutes} min `;
    }
    if (hours === 0 && minutes === 0 && seconds > 0) {
        result += `${seconds} sec `;
    }
    return result.substr(0, result.length - 1);
};

module.exports = { setPageContentHeight, toggleFooterPosition, fetchLocationInfo, updateLocaleLanguagesDropdown ,calculateTime, formatTime, getLocaleString, performAPIRequest}
