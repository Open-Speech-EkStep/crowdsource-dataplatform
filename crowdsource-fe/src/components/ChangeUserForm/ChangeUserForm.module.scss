.root {
  font-size: 0.875rem;
  line-height: 1.125rem;
}

.radio label {
  line-height: 1.375rem;
}
