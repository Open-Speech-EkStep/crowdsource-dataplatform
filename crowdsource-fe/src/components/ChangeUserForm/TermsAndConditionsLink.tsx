import { useTranslation } from 'next-i18next';

import Link from 'components/Link';
import routePaths from 'constants/routePaths';

import styles from './ChangeUserForm.module.scss';

const TermsAndConditionsLink = () => {
  const { t } = useTranslation();

  return (
    <Link href={routePaths.termsAndConditions}>
      <a target="_blank" className={styles.termsAndConditionsLink}>
        {t('termsAndConditionsTextValue')}
      </a>
    </Link>
  );
};

export default TermsAndConditionsLink;
