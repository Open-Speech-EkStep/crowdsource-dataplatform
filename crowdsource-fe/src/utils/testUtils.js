import { render } from '@testing-library/react';

const Wrapper = ({ children }) => children;

const customRender = (ui, options) => render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };
