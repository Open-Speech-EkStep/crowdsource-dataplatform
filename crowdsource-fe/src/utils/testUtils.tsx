/* eslint-disable jest/no-export */
import type { ReactNode, ReactElement, ComponentType } from 'react';

import type { RenderOptions, RenderResult } from '@testing-library/react';
import { render, act } from '@testing-library/react';
import { axe } from 'jest-axe';
import puppeteer from 'puppeteer';
import type { Browser } from 'puppeteer';

const Wrapper: ComponentType = ({ children }: { children?: ReactNode }) => children as ReactElement;

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };

export const verifyAxeTest = ({ container }: { container: RenderResult['container'] }) => {
  it('should not fail an axe audit', async () => {
    let axeResult;

    await act(async () => {
      axeResult = await axe(container);
    });

    expect(axeResult).toHaveNoViolations();
  });
};

export const verifyVRTest = (title: string, url: string) => {
  describe(`${title} VR test`, () => {
    let browser: Browser;

    beforeAll(async () => {
      browser = await puppeteer.launch({
        headless: true,
        devtools: false,
        ignoreHTTPSErrors: true,
        args: ['--start-fullscreen'],
      });
    });

    it('match snapshot', async () => {
      const page = await browser.newPage();
      await page.setViewport({ width: 1920, height: 1080 });

      await page.goto(url, {
        waitUntil: 'networkidle2',
      });

      await page.addStyleTag({
        content: `
      *,
      *::after,
      *::before {
          transition-delay: 0s !important;
          transition-duration: 0s !important;
          animation-delay: -0.0001s !important;
          animation-duration: 0s !important;
          animation-play-state: paused !important;
          caret-color: transparent !important;
      }
      `,
      });

      const image = await page.screenshot({
        fullPage: true,
      });

      expect(image).toMatchImageSnapshot();
    });

    afterAll(async () => {
      await browser.close();
    });
  });
};
