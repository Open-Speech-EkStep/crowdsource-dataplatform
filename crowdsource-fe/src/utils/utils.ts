import { jsPDF } from 'jspdf';
import { i18n } from 'next-i18next';

import { ErrorStatusCode } from 'constants/errorStatusCode';
import { INITIATIVES_MAPPING } from 'constants/initiativeConstants';
import { KEYBOARD_ERROR, LANGUAGE_UNICODE, OTHER_LANGUAGE_UNICODE } from 'constants/Keyboard';
import type { InitiativeType } from 'types/InitiativeType';

import apiPaths from '../constants/apiPaths';

var platform = require('platform');

export const convertIntoHrsFormat = (data: any, isSeconds = true) => {
  const hours = Math.floor(data / 3600);
  const remainingAfterHours = data % 3600;
  const minutes = Math.floor(remainingAfterHours / 60);
  const seconds = Math.round(remainingAfterHours % 60);
  if (isSeconds) {
    return { hours, minutes, seconds };
  } else {
    return { hours, minutes };
  }
};

export const roundOffValue = (average: any, precision: number) => {
  const formattedAverage = average.toFixed(precision);
  return formattedAverage;
};

export const capitalizeFirstLetter = (strValue: string) => {
  return strValue?.charAt(0).toUpperCase() + strValue?.slice(1);
};

const translateText = (text: string) => i18n?.t(text);

export const formatTime = (hours: number, minutes = 0, seconds = 0) => {
  const hrStr = translateText('hours1');
  const minStr = translateText('minutes1');
  const secStr = translateText('seconds1');
  let result = '';
  if (hours > 0) {
    result += `${hours} ${hrStr} `;
  }
  if (minutes > 0) {
    result += `${minutes} ${minStr} `;
  }
  if (hours === 0 && minutes === 0 && seconds > 0) {
    result += `${seconds} ${secStr} `;
  }

  if (hours === 0 && minutes === 0 && seconds === 0) {
    result += `0 ${secStr} `;
  }

  return result.substr(0, result.length - 1);
};

export const convertTimeFormat = (value: any) => {
  const { hours, minutes, seconds } = convertIntoHrsFormat(Number(value) * 60 * 60);
  const data = formatTime(hours, minutes, seconds);
  return data;
};

const roundDuration = (value: number) => {
  return Math.round(value * 100) / 100;
};

export const getHoursValue = (value: any) => {
  return roundDuration(value);
};

export const getMinutesValue = (value: any) => {
  return roundDuration(value * 60);
};

export const getHoursText = (value: any) => {
  return `${getHoursValue(value)} ${i18n?.t('hours2')}`;
};

export const getMinutesText = (value: any) => {
  return `${getMinutesValue(value)} ${i18n?.t('minutes2')}`;
};

export const isSunoOrBoloInitiative = (value: InitiativeType) => {
  const boloOrSuno = ['asr', 'text'];
  return boloOrSuno.includes(value);
};

export const isBoloInitiative = (value: InitiativeType) => {
  const boloOrSuno = ['text'];
  return boloOrSuno.includes(value);
};

export const findInputError = (text: string, currentModule: string, language: string) => {
  if (!text.length) return;
  const newText = text.replace(/\s/g, ''); //read input value, and remove "space" by replace \s
  let error: any = KEYBOARD_ERROR.language;
  const isSunoModule = currentModule === INITIATIVES_MAPPING.suno;
  const specialSymbols = /[\u0021-\u002F\u003A-\u0040\u005B-\u0060\u007B-\u007E\u0964-\u0965]/;
  if (currentModule === INITIATIVES_MAPPING.suno && specialSymbols.test(newText) === true) {
    return KEYBOARD_ERROR.symbol;
  }
  Object.entries(isSunoModule ? LANGUAGE_UNICODE : OTHER_LANGUAGE_UNICODE).forEach(([key, value]) => {
    if (value.test(newText)) {
      //Check Unicode to see which one is true
      if (key.toLowerCase() === language.toLowerCase()) {
        error = KEYBOARD_ERROR.noError;
      }
    }
  });
  return error;
};

export const fetchLocationInfo = async () => {
  const result = await fetch('https://www.cloudflare.com/cdn-cgi/trace')
    .then(res => res.text())
    .then(async ipAddressText => {
      const dataArray = ipAddressText.split('\n');
      let ipAddress = '';
      for (let ind in dataArray) {
        if (dataArray[ind].startsWith('ip=')) {
          ipAddress = dataArray[ind].replace('ip=', '');
          break;
        }
      }
      if (ipAddress.length !== 0) {
        const data = await fetch(`${apiPaths.locationInfo}?ip=${ipAddress}`);
        return data && data.json();
      } else {
        return new Promise((resolve, reject) => {
          reject('Ip Address not available');
        });
      }
    });

  return result;
};

export const getDeviceInfo = () => {
  const os = platform.os;
  return os.family + ' ' + os.version;
};

export const getBrowserInfo = () => {
  return platform.name + ' ' + platform.version;
};

export const downloadBadge = (
  locale: string | undefined,
  initiative: string,
  source: string,
  badgeType: string,
  badgeId: string
) => {
  const pdf = new jsPDF();
  const img = new Image();
  img.src = `/images/${locale}/badges/${locale}_${initiative}_${badgeType}_${source}.png`;
  img.crossOrigin = 'Anonymous';
  img.onload = function () {
    pdf.addImage(img, 50, 10, 105, 130);
    pdf.save(`${badgeType}-badge.pdf`);
  };

  pdf.text(`Badge Id : ${badgeId}`, 36, 190);
};

export const groupBy = function (list: any, keyName: any) {
  if (!list || list.length === 0 || !list[0][keyName]) {
    return [];
  }
  return list.reduce(function (pair: any, key: any) {
    (pair[key[keyName]] = pair[key[keyName]] || []).push(key);
    return pair;
  }, {});
};

/* istanbul ignore next */
export const visualize = (visualizer: any, analyser: any) => {
  const canvasCtx = visualizer.getContext('2d');
  const bufferLength = analyser.frequencyBinCount;
  const dataArray = new Uint8Array(bufferLength);
  const WIDTH = visualizer.width;
  const HEIGHT = visualizer.height;
  // TODO do we need to limit the number of time visualize refreshes per second
  // so that it can run on Android processors without causing audio to drop?
  function draw() {
    // this is more efficient than calling with processor.onaudioprocess
    // and sending floatarray with each call...
    requestAnimationFrame(draw);
    analyser.getByteTimeDomainData(dataArray);
    // canvasCtx.fillStyle = 'rgb(255, 255, 255, 0)';
    // canvasCtx.fillRect(0, 0, WIDTH, HEIGHT);
    canvasCtx.clearRect(0, 0, WIDTH, HEIGHT);
    canvasCtx.lineWidth = 2;
    canvasCtx.strokeStyle = 'rgb(0,123,255)';
    canvasCtx.beginPath();
    const sliceWidth = (WIDTH * 1.0) / bufferLength;
    let x = 0;
    for (let i = 0; i < bufferLength; i++) {
      let v = dataArray[i] / 128.0; // uint8
      let y = (v * HEIGHT) / 2; // uint8
      if (i === 0) {
        canvasCtx.moveTo(x, y);
      } else {
        canvasCtx.lineTo(x, y);
      }
      x += sliceWidth;
    }
    canvasCtx.lineTo(visualizer.width, visualizer.height / 2);
    canvasCtx.stroke();
  }

  draw();
};

export const getErrorMsg = (error: any) => {
  if (error && error.status === ErrorStatusCode.TOO_MANY_REQUEST) {
    return 'multipleRequestApiError';
  } else {
    return 'apiFailureError';
  }
};
