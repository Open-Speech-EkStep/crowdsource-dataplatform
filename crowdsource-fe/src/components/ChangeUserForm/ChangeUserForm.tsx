import { useState } from 'react';
import type { ChangeEvent, FormEvent } from 'react';

import { useTranslation, Trans } from 'next-i18next';
import { useRouter } from 'next/router';
import Form from 'react-bootstrap/Form';

import { DISPLAY_LANGUAGES, RAW_LANGUAGES, DEFAULT_LOCALE } from 'constants/localesConstants';
import localStorageConstants from 'constants/localStorageConstants';
import useLocalStorage from 'hooks/useLocalStorage';

import styles from './ChangeUserForm.module.scss';
import TermsAndConditionsLink from './TermsAndConditionsLink';

const testUserName = (value: string) => {
  const mobileRegex = /^[6-9]\d{9}$/;
  const emailRegex = /^\S+@\S+[.][0-9a-z]+$/;

  return mobileRegex.test(value) || emailRegex.test(value);
};

interface ChangeUserFormProps {
  onSubmit: () => void;
}

const ChangeUserForm = ({ onSubmit }: ChangeUserFormProps) => {
  const { locales, locale: currentLocale = DEFAULT_LOCALE } = useRouter();
  const { t } = useTranslation();
  const [, setSpeakerDetails] = useLocalStorage(localStorageConstants.speakerDetails);

  const [isOtherGenderVisible, setOtherGenderVisibility] = useState(false);
  const [formData, setFormData] = useState({
    userName: '',
    motherTongue: '',
    age: '',
    gender: '',
    otherGender: '',
  });

  const hideOtherGender = () => {
    setOtherGenderVisibility(false);
  };
  const showOtherGender = () => {
    setOtherGenderVisibility(true);
  };

  const isUserNameInValid = testUserName(formData.userName);
  const isFormInValid = isUserNameInValid;

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    setFormData({ ...formData, [e.target.name]: e.target.value });
  };
  const handleGenderChange = (e: ChangeEvent<HTMLInputElement>) => {
    hideOtherGender();
    setFormData({ ...formData, [e.target.name]: e.target.value, otherGender: '' });
  };
  const handleOtherGenderChange = () => {
    showOtherGender();
    setFormData({ ...formData, otherGender: 'Rather Not Say' });
  };

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (!isFormInValid) {
      setSpeakerDetails({
        ...formData,
        gender: formData.otherGender || formData.gender,
        otherGender: undefined,
        language: RAW_LANGUAGES[currentLocale],
        toLanguage: '',
      });

      onSubmit();
    }
  };

  return (
    <Form
      data-testid="ChangeUserForm"
      id="changeUserForm"
      className={`${styles.root} py-2`}
      noValidate
      onSubmit={handleSubmit}
    >
      <Form.Group className="py-3" controlId="userName">
        <Form.Label>{t('name')}</Form.Label>
        <Form.Control
          type="text"
          name="userName"
          placeholder={t('enterUsername')}
          maxLength={12}
          isInvalid={isUserNameInValid}
          value={formData.userName}
          onChange={handleChange}
        />
        {isUserNameInValid && (
          <Form.Control.Feedback type="invalid">{t('userNameError')}</Form.Control.Feedback>
        )}
        {!isUserNameInValid && <Form.Text className="d-block text-muted mt-1">{t('userNameHint')}</Form.Text>}
      </Form.Group>

      <Form.Group className="py-3" controlId="motherTongue">
        <Form.Label className="mb-1">{t('motherTongue')}</Form.Label>
        <Form.Select
          name="motherTongue"
          className={styles.select}
          value={formData.motherTongue}
          onChange={handleChange as any}
        >
          <option value="">{t('selectMotherTongue')}</option>
          {locales?.map(locale => (
            <option key={locale} value={RAW_LANGUAGES[locale]}>
              {t(DISPLAY_LANGUAGES[locale])}
            </option>
          ))}
        </Form.Select>
      </Form.Group>

      <Form.Group className="py-3" controlId="age">
        <Form.Label className="mb-1">{t('ageGroup')}</Form.Label>
        <Form.Select name="age" className={styles.select} value={formData.age} onChange={handleChange as any}>
          <option value="">{t('selectAgeGroup')}</option>
          <option value="upto 10">{t('kidAgeGroup')}</option>
          <option value="10 - 30">{t('youthAgeGroup')}</option>
          <option value="30 - 60">{t('adultAgeGroup')}</option>
          <option value="60+">{t('seniorAgeGroup')}</option>
        </Form.Select>
      </Form.Group>

      <Form.Group className="py-3" controlId="gender">
        <Form.Label className="mb-1">{t('gender')}</Form.Label>
        <div>
          <Form.Check
            inline
            type="radio"
            id="male"
            label={t('male')}
            name="gender"
            className={`${styles.radio} me-8 mb-0`}
            onChange={handleGenderChange}
            value="male"
          />
          <Form.Check
            inline
            type="radio"
            id="female"
            label={t('female')}
            name="gender"
            className={`${styles.radio} me-8 mb-0`}
            onChange={handleGenderChange}
            value="female"
          />
          <Form.Check
            inline
            type="radio"
            id="others"
            label={t('others')}
            className={`${styles.radio} me-8 mb-0`}
            name="gender"
            onChange={handleOtherGenderChange}
            value=""
          />
        </div>
        {isOtherGenderVisible && (
          <div>
            <Form.Check
              inline
              type="radio"
              id="transgenderMale"
              label={t('transgenderMale')}
              name="otherGender"
              className={`${styles.radio} me-8 mb-0`}
              onChange={handleChange}
              value="Transgender - He"
            />
            <Form.Check
              inline
              type="radio"
              id="transgenderFemale"
              label={t('transgenderFemale')}
              name="otherGender"
              className={`${styles.radio} me-8 mb-0`}
              onChange={handleChange}
              value="Transgender - She"
            />
            <Form.Check
              inline
              type="radio"
              id="ratherNotSayGender"
              label={t('ratherNotSayGender')}
              name="otherGender"
              className={`${styles.radio} me-8 mb-0`}
              onChange={handleChange}
              value="Rather Not Say"
              defaultChecked
            />
          </div>
        )}
      </Form.Group>
      <p className="py-3 mb-0">
        <Trans
          i18nKey="termsConditionsText"
          defaults="termsConditionsText"
          components={{
            TermsAndConditionsLink: <TermsAndConditionsLink />,
          }}
        />
      </p>
    </Form>
  );
};

export default ChangeUserForm;
