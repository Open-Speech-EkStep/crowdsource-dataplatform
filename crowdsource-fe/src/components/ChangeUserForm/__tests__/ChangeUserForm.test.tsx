import router from 'next/router';

import { render, verifyAxeTest, screen, userEvent } from 'utils/testUtils';

import ChangeUserForm from '../ChangeUserForm';

describe('ChangeUserForm', () => {
  const setup = (redirectionUrl?: any, doRedirection?: boolean) => {
    router.locale = undefined;
    const onSubmit = jest.fn();

    return {
      ...render(
        <>
          <ChangeUserForm onSubmit={onSubmit} doRedirection={doRedirection} redirectionUrl={redirectionUrl} />
          <button form="changeUserForm">submit</button>
        </>
      ),
      onSubmit,
    };
  };

  verifyAxeTest(setup());

  it('should render the component and matches it against stored snapshot', () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should handle for valid/invalid username', () => {
    setup('/suno-india', true);

    expect(screen.queryByText('userNameError')).not.toBeInTheDocument();
    expect(screen.getByText('userNameHint')).toBeInTheDocument();

    userEvent.clear(screen.getByRole('textbox', { name: 'name' }));
    userEvent.type(screen.getByRole('textbox', { name: 'name' }), 'abc@abc.com');

    expect(screen.getByText('userNameError')).toBeInTheDocument();
    expect(screen.queryByText('userNameHint')).not.toBeInTheDocument();

    userEvent.clear(screen.getByRole('textbox', { name: 'name' }));
    userEvent.type(screen.getByRole('textbox', { name: 'name' }), '9879879878');

    expect(screen.getByText('userNameError')).toBeInTheDocument();
    expect(screen.queryByText('userNameHint')).not.toBeInTheDocument();

    userEvent.clear(screen.getByRole('textbox', { name: 'name' }));
    userEvent.type(screen.getByRole('textbox', { name: 'name' }), 'abc_123');

    expect(screen.queryByText('userNameError')).not.toBeInTheDocument();
    expect(screen.getByText('userNameHint')).toBeInTheDocument();
  });

  it('should save user details in localstorage', () => {
    const { onSubmit } = setup();

    userEvent.clear(screen.getByRole('textbox', { name: 'name' }));
    userEvent.type(screen.getByRole('textbox', { name: 'name' }), 'abc@abc.com');

    userEvent.click(screen.getByRole('button', { name: 'submit' }));

    expect(localStorage.setItem).not.toHaveBeenCalledTimes(1);
    expect(onSubmit).not.toHaveBeenCalledTimes(1);

    userEvent.clear(screen.getByRole('textbox', { name: 'name' }));
    userEvent.type(screen.getByRole('textbox', { name: 'name' }), 'abc');

    userEvent.click(screen.getByRole('button', { name: 'submit' }));

    expect(localStorage.setItem).toHaveBeenCalledWith(
      'speakerDetails',
      '{"userName":"abc","motherTongue":"","age":"","gender":"","language":"English","toLanguage":""}'
    );
    expect(onSubmit).toHaveBeenCalledTimes(1);

    userEvent.click(screen.getByRole('combobox', { name: 'motherTongue' }));
    userEvent.selectOptions(screen.getByRole('combobox', { name: 'motherTongue' }), 'Hindi');

    userEvent.click(screen.getByRole('combobox', { name: 'ageGroup' }));
    userEvent.selectOptions(screen.getByRole('combobox', { name: 'ageGroup' }), '30 - 60');

    userEvent.click(screen.getByRole('radio', { name: 'male' }));

    jest.clearAllMocks();

    userEvent.click(screen.getByRole('button', { name: 'submit' }));

    expect(localStorage.setItem).toHaveBeenCalledWith(
      'speakerDetails',
      '{"userName":"abc","motherTongue":"Hindi","age":"30 - 60","gender":"male","language":"English","toLanguage":""}'
    );
    expect(onSubmit).toHaveBeenCalledTimes(1);
  });

  it('should handle gender radio combinations', () => {
    setup(null, true);

    userEvent.click(screen.getByRole('radio', { name: 'male' }));

    expect(screen.getByRole('radio', { name: 'male' })).toBeChecked();
    expect(screen.queryByRole('radio', { name: 'transgenderMale' })).not.toBeInTheDocument();
    expect(screen.queryByRole('radio', { name: 'transgenderFemale' })).not.toBeInTheDocument();
    expect(screen.queryByRole('radio', { name: 'ratherNotSayGender' })).not.toBeInTheDocument();

    userEvent.click(screen.getByRole('radio', { name: 'female' }));

    expect(screen.getByRole('radio', { name: 'male' })).not.toBeChecked();
    expect(screen.getByRole('radio', { name: 'female' })).toBeChecked();
    expect(screen.queryByRole('radio', { name: 'transgenderMale' })).not.toBeInTheDocument();
    expect(screen.queryByRole('radio', { name: 'transgenderFemale' })).not.toBeInTheDocument();
    expect(screen.queryByRole('radio', { name: 'ratherNotSayGender' })).not.toBeInTheDocument();

    userEvent.click(screen.getByRole('radio', { name: 'others' }));

    expect(screen.getByRole('radio', { name: 'male' })).not.toBeChecked();
    expect(screen.getByRole('radio', { name: 'female' })).not.toBeChecked();
    expect(screen.getByRole('radio', { name: 'others' })).toBeChecked();
    expect(screen.getByRole('radio', { name: 'transgenderMale' })).toBeInTheDocument();
    expect(screen.getByRole('radio', { name: 'transgenderFemale' })).toBeInTheDocument();
    expect(screen.getByRole('radio', { name: 'ratherNotSayGender' })).toBeInTheDocument();
    expect(screen.getByRole('radio', { name: 'ratherNotSayGender' })).toBeChecked();

    jest.clearAllMocks();
    userEvent.click(screen.getByRole('button', { name: 'submit' }));
    expect(localStorage.setItem).toHaveBeenCalledWith(
      'speakerDetails',
      '{"userName":"","motherTongue":"","age":"","gender":"Rather Not Say","language":"English","toLanguage":""}'
    );

    userEvent.click(screen.getByRole('radio', { name: 'transgenderMale' }));

    jest.clearAllMocks();
    userEvent.click(screen.getByRole('button', { name: 'submit' }));
    expect(localStorage.setItem).toHaveBeenCalledWith(
      'speakerDetails',
      '{"userName":"","motherTongue":"","age":"","gender":"Transgender - He","language":"English","toLanguage":""}'
    );

    userEvent.click(screen.getByRole('radio', { name: 'female' }));

    expect(screen.getByRole('radio', { name: 'female' })).toBeChecked();
    expect(screen.queryByRole('radio', { name: 'transgenderMale' })).not.toBeInTheDocument();
    expect(screen.queryByRole('radio', { name: 'transgenderFemale' })).not.toBeInTheDocument();
    expect(screen.queryByRole('radio', { name: 'ratherNotSayGender' })).not.toBeInTheDocument();

    jest.clearAllMocks();
    userEvent.click(screen.getByRole('button', { name: 'submit' }));
    expect(localStorage.setItem).toHaveBeenCalledWith(
      'speakerDetails',
      '{"userName":"","motherTongue":"","age":"","gender":"female","language":"English","toLanguage":""}'
    );
  });
});
