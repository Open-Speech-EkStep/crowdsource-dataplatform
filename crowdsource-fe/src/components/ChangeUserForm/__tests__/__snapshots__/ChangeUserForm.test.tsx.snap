// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeUserForm should render the component and matches it against stored snapshot 1`] = `
<DocumentFragment>
  <form
    class="root py-2"
    data-testid="ChangeUserForm"
    id="changeUserForm"
    novalidate=""
  >
    <div
      class="py-3"
    >
      <label
        class="form-label"
        for="userName"
      >
        name
      </label>
      <input
        class="form-control"
        id="userName"
        maxlength="12"
        name="userName"
        placeholder="enterUsername"
        type="text"
        value=""
      />
      <small
        class="d-block text-muted mt-1 form-text"
      >
        userNameHint
      </small>
    </div>
    <div
      class="py-3"
    >
      <label
        class="mb-1 form-label"
        for="motherTongue"
      >
        motherTongue
      </label>
      <select
        class="select form-select"
        id="motherTongue"
        name="motherTongue"
      >
        <option
          value=""
        >
          selectMotherTongue
        </option>
        <option
          value="English"
        >
          English
        </option>
        <option
          value="Hindi"
        >
          हिंदी
        </option>
        <option
          value="Tamil"
        >
          தமிழ்
        </option>
        <option
          value="Telugu"
        >
          తెలుగు
        </option>
        <option
          value="Assamese"
        >
          অসমীয়া
        </option>
        <option
          value="Bengali"
        >
          বাংলা
        </option>
        <option
          value="Gujarati"
        >
          ગુજરાતી
        </option>
        <option
          value="Kannada"
        >
          ಕನ್ನಡ
        </option>
        <option
          value="Malayalam"
        >
          മലയാളം
        </option>
        <option
          value="Marathi"
        >
          मराठी
        </option>
        <option
          value="Odia"
        >
          ଓଡିଆ
        </option>
        <option
          value="Punjabi"
        >
          ਪੰਜਾਬੀ
        </option>
      </select>
    </div>
    <div
      class="py-3"
    >
      <label
        class="mb-1 form-label"
        for="age"
      >
        ageGroup
      </label>
      <select
        class="select form-select"
        id="age"
        name="age"
      >
        <option
          value=""
        >
          selectAgeGroup
        </option>
        <option
          value="upto 10"
        >
          kidAgeGroup
        </option>
        <option
          value="10 - 30"
        >
          youthAgeGroup
        </option>
        <option
          value="30 - 60"
        >
          adultAgeGroup
        </option>
        <option
          value="60+"
        >
          seniorAgeGroup
        </option>
      </select>
    </div>
    <div
      class="py-3"
    >
      <label
        class="mb-1 form-label"
        for="gender"
      >
        gender
      </label>
      <div>
        <div
          class="radio me-8 mb-0 form-check form-check-inline"
        >
          <input
            class="form-check-input"
            id="male"
            name="gender"
            type="radio"
            value="male"
          />
          <label
            class="form-check-label"
            for="male"
            title=""
          >
            male
          </label>
        </div>
        <div
          class="radio me-8 mb-0 form-check form-check-inline"
        >
          <input
            class="form-check-input"
            id="female"
            name="gender"
            type="radio"
            value="female"
          />
          <label
            class="form-check-label"
            for="female"
            title=""
          >
            female
          </label>
        </div>
        <div
          class="radio me-8 mb-0 form-check form-check-inline"
        >
          <input
            class="form-check-input"
            id="others"
            name="gender"
            type="radio"
            value=""
          />
          <label
            class="form-check-label"
            for="others"
            title=""
          >
            others
          </label>
        </div>
      </div>
    </div>
    <p
      class="py-3 mb-0"
    >
      termsConditionsText
    </p>
  </form>
  <button
    form="changeUserForm"
  >
    submit
  </button>
</DocumentFragment>
`;
